prometheus_hardware_exporter/collectors/redfish.py

```diff
@@ -56,23 +86,338 @@ def _retrieve_redfish_sensor_data(self) -> List[Any]:
         Returns:
             sensors: List of dicts with details for each sensor
         """
-        try:
-            with redfish.redfish_client(
-                base_url=self.host,
-                username=self.username,
-                password=self.password,
-                timeout=self.timeout,
-                max_retry=self.max_retry,
-            ) as redfish_obj:
-                return redfish_utilities.get_sensors(redfish_obj)
-        except (
-            InvalidCredentialsError,
-            SessionCreationError,
-            ConnectionError,
-            RetriesExhaustedError,
-        ) as err:
-            logger.exception(err)
-        return []
+        sensors: List[Any] = []
+
+        redfish_obj: Optional[HttpClient] = self._get_redfish_obj()
```

Comment:
**issue (blocking):**  

I saw this `_get_redfish_obj` in multiple different fetch data function.
I would like to know how long this function takes. Because multiple times login/logout maybe cause performance issue here.

Also I would like to keep the login session for later use. Can we use some similar design pattern from web crawler?

Reply:
Here is a [pastebin link](https://pastebin.ubuntu.com/p/tmwhJB6cJQ/) to the full output when querying from the exporter using only the redfish collector.

- The average time for running `curl localhost:10000` currently is about 11s. 
- Creating a new REST session and logging in takes about 0.8s. Logging out takes about 0.1s. So the whole process would be around 1s.
- We're creating a new redfish object and performing login/logout for each metric, or 8 times or about 8s. If we keep only one session to use for each metric and login/logout only once, we can save 7s here.
- Most of the other API calls take < 0.1s

I would need to see how I can instantiate a new redfish object once when the `RedfishHelper` class object is created  and then logout once this object is destroyed.

Reply:
So we can ask some questions here:

- How long can I keep this session open?
- When should I create the session?
  - Should I create the session when the service start?
  - Should I create the session when I receive the request?
- Where should I store this session?
- Is there any pattern that can help me to get the same session every time when I send the request to redfish endpoint?

Reply:
This is now resolved since I've added `__enter__` and `__exit__` methods to RedfishHelper class to allow it to be used as a context manager. Now when the context manager exists, the `logout` method is called and the session is closed. Also, we only have to login/logout once now.